spring-kafka/src/main/java/org/springframework/kafka/listener/BatchListenerFailedException.java

```diff
@@ -0,0 +1,69 @@
+/*
+ * Copyright 2020 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.kafka.listener;
+
+import org.springframework.kafka.KafkaException;
+import org.springframework.lang.Nullable;
+
+/**
+ * An exception thrown by user code to inform the framework which record in a batch has
+ * failed.
+ *
+ * @author Gary Russell
+ * @since 2.5
+ *
+ */
+public class BatchListenerFailedException extends KafkaException {
+
+	private static final long serialVersionUID = 1L;
+
+	private final int index;
+
+	/**
+	 * Construct an instance with the provided properties.
+	 * @param message the message.
+	 * @param index the index in the batch of the failed record.
+	 */
+	public BatchListenerFailedException(String message, int index) {
```

Comment:
I wonder why it can't be a `ConsumerRecord` per se? 
Why should I worry about record selection in the batch by myself?

Reply:
Good idea

Reply:
However, it's no good when using a POJO listener - I can come up with a hybrid.